### PR TITLE
feat: add mindspore-version CLIP (rewritten based on openAI-CLIP)

### PR DESCRIPTION
**Pull request by XixinYang**
Thank you for your contribution to the MindCV repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

Add mindspore-version CLIP which is totally rewritten based on openAI-CLIP.

## Test Plan

These code have been **tested on the first example in ./README.md with both ViT and ResNet ckpts**, which has used all the main APIS, and the result differences are as below:
ViT:
image
ResNet:
image
**After discussion, all the diffences are considered within an acceptable range.**

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

**Comment by SamitHuang** [comment]
As a  text-image alignment foundation mode, CLIP is not appropriate to place in `examples`. We can put it under `mindone`

**Comment by XixinYang** [comment]
> As a text-image alignment foundation mode, CLIP is not appropriate to place in `examples`. We can put it under `mindone`

**Comment by SamitHuang** [comment]
> As a text-image alignment foundation mode, CLIP is not appropriate to place in `examples`. We can put it under `mindone`

**Comment by SamitHuang** [comment]
> As a text-image alignment foundation mode, CLIP is not appropriate to place in `examples`. We can put it under `mindone`

Take back by words. Let MindCV be its root.

File: examples/clip/clip/ckpt_transform.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        elif ".bn" in name or ".downsample.1." in name:
+            name = name.replace(".weight", ".gamma").replace(".bias", ".beta")
+            name = name.replace(".running_mean", ".moving_mean").replace(".running_var", ".moving_variance")
+        pt_params.append({"name": name, "data": ms.Tensor(parameter.numpy())})
+    return pt_params
+
+
+pth_path = "./ViT-B-32.pt"
+pt_param = pytorch_params(pth_path)
+ms.save_checkpoint(pt_param, "./ViT-B-32.ckpt")
```
Comment (SamitHuang) [inline examples/clip/clip/ckpt_transform.py:24]: It's better to put them under `if __name__ == "__main__", and allow input args for different clip model version, e..g vit-h

File: examples/clip/clip/clip.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    BICUBIC = Image.BICUBIC
+
+if packaging.version.parse(ms.__version__) < packaging.version.parse("2.0.0"):
+    warnings.warn("MindSpore version 2.0.0 or higher is recommended")
+
+__all__ = ["available_models", "load", "tokenize"]
+_tokenizer = _Tokenizer()
+
+_MODELS = {
+    "RN50": "https://openaipublic.azureedge.net/clip/models/afeb0e10f9e5a86da6080e35cf09123aca3b358a0c3e3b6c78a7b63bc04b6762/RN50.pt",
```
Comment (SamitHuang) [inline examples/clip/clip/clip.py:33]: We should use mindspore *.ckpt by default, which can be updated after uploading the converted weights to download servers.

File: examples/clip/clip/ckpt_transform.py (hunk 1 above)
Comment (XixinYang) [reply examples/clip/clip/ckpt_transform.py]: Great idea and have done now! ✅

File: examples/clip/README.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
+# CLIP
+
+[[Blog]](https://openai.com/blog/clip/) [[Paper]](https://arxiv.org/abs/2103.00020) [[Model Card]](model-card.md) [[Colab]](https://colab.research.google.com/github/openai/clip/blob/master/notebooks/Interacting_with_CLIP.ipynb)[[Source Code]](https://github.com/openai/CLIP)
```
Comment (geniuspatrick) [inline examples/clip/README.md:3]: 链接

File: examples/clip/README.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+![CLIP](CLIP.png)
+
+
+## Usage
+
+First, [install MindSpore 2.0.0](https://www.mindspore.cn/install) (or later), as well as small additional dependencies.
+
+```bash
+$ pip install ftfy regex tqdm mindspore
```
Comment (geniuspatrick) [inline examples/clip/README.md:19]: 为什么不用requirements.txt文件？去掉mindspore

File: examples/clip/README.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+$ cd ./examples/clip/
+```
+
+Mindspore does not support .pt/.pth file, you are strongly recommended to transform the checkpoint file from OpenAI to a .ckpt format. For example:
+
+```bash
+$ python ./clip/ckpt_transform.py --pth_path="ViT-B-32"
```
Comment (geniuspatrick) [inline examples/clip/README.md:26]: 这个文件最好放在外面

File: examples/clip/README.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+### Zero-Shot Prediction
+
+The code below performs zero-shot prediction using CLIP, as shown in Appendix B in the paper. This example takes an image from the [CIFAR-100 dataset](https://www.cs.toronto.edu/~kriz/cifar.html) (which has been integrated by MindSpore and hence we download it from MindSpore in this example), and predicts the most likely labels among the 100 textual labels from the dataset.
+
+```python
+import clip
+from mindspore import ops, nn, Tensor
+import mindspore.dataset as ds
+from download import download
```
Comment (geniuspatrick) [inline examples/clip/README.md:97]: 这是个啥包

File: examples/clip/README.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+accuracy = np.mean((test_labels == predictions).astype(float)) * 100.
+print(f"Accuracy = {accuracy:.3f}")
+```
+
+Note that the `C` value should be determined via a hyperparameter sweep using a validation split.
+
+
+## See Also
+
+* [OpenCLIP](https://github.com/mlfoundations/open_clip): includes larger and independently trained CLIP models up to ViT-G/14
```
Comment (geniuspatrick) [inline examples/clip/README.md:207]: 这个记得到时候改成我们的链接

**Comment by XixinYang** [comment]
TODO: add an API check list in a susquent PR.

**Review by geniuspatrick — APPROVED** [review]

File: examples/clip/README.md [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+test_features, test_labels = get_features(cifar100_test_iter)
+
+# Perform logistic regression
+classifier = LogisticRegression(random_state=0, C=0.316, max_iter=1000, verbose=1)
+classifier.fit(train_features, train_labels)
+
+# Evaluate using the logistic regression classifier
+predictions = classifier.predict(test_features)
+accuracy = np.mean((test_labels == predictions).astype(float)) * 100.
+print(f"Accuracy = {accuracy:.3f}")
```
Comment (SamitHuang) [inline examples/clip/README.md:219]: Suggest to show the accuracy of linear-probe and zero-shot image classification, which will finally prove the precision and correctness of our implementation

**Review by SamitHuang — APPROVED** [review]

File: examples/clip/README.md (hunk 8 above)
Comment (SamitHuang) [reply examples/clip/README.md:219]: Compare with the original CLIP paper,  how is the zero-shot and linear-probe ACC?

**Comment by XixinYang** [comment]
TODO：Compare with the original CLIP paper on the zero-shot and linear-probe ACC (imagenet)

**Comment by XixinYang** [comment]
> TODO: add an API check list in a susquent PR.

done
